packages/story-editor/src/app/rightClickMenu/hooks/useIsPartiallyOffCanvas.js

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/**
+ * External dependencies
+ */
+import {
+  DANGER_ZONE_HEIGHT,
+  PAGE_HEIGHT,
+  PAGE_WIDTH,
+} from '@googleforcreators/units';
+
+function useIsPartiallyOffCanvas(selectedElement) {
+  const { x, y, width, height } = selectedElement;
+  let isOffCanvas = false;
+  let offCanvasTop = 0;
+  let offCanvasRight = 0;
+  let offCanvasBottom = 0;
+  let offCanvasLeft = 0;
+
+  // if x is a negative #
+  if (x < 0) {
+    offCanvasLeft = Math.abs(x);
+  }
+
+  // We have to adjust the y value since y = 0 is the top edge of the page container, not the top edge of the fullbleed container.
+  const adjustedY = y + DANGER_ZONE_HEIGHT;
+  // if y is a negative #
+  if (adjustedY < 0) {
+    offCanvasTop = Math.abs(adjustedY);
+  }
+
+  // check for off-canvas right side
+  if (x > 0) {
+    if (x + width > PAGE_WIDTH) {
+      offCanvasRight = x + width - PAGE_WIDTH;
+    }
+  } else {
+    const r = width - Math.abs(x) - PAGE_WIDTH;
+    offCanvasRight = r > 0 ? r : 0;
+  }
+
+  // check if off-canvas bottom
+  if (adjustedY > 0) {
+    if (adjustedY + height > PAGE_HEIGHT) {
+      offCanvasBottom = adjustedY + height - PAGE_HEIGHT;
+    }
+  } else {
+    const b = height - Math.abs(y) - PAGE_HEIGHT;
+    offCanvasBottom = b > 0 ? b : 0;
+  }
+
+  if (
+    offCanvasTop > 0 ||
+    offCanvasRight > 0 ||
+    offCanvasBottom > 0 ||
+    offCanvasLeft > 0
+  ) {
+    isOffCanvas = true;
+  }
+
+  const percentage = width / selectedElement.resource.width;
+  const multiplier = 100 / (percentage * 100);
+  const offCanvasX = Math.floor((offCanvasLeft + offCanvasRight) * multiplier);
+  const offCanvasY = Math.floor((offCanvasTop + offCanvasBottom) * multiplier);
+  const cropWidth = selectedElement.resource.width - offCanvasX;
+  const cropHeight = selectedElement.resource.height - offCanvasY;
+  const cropX = Math.floor(offCanvasLeft * multiplier);
+  const cropY = Math.floor(offCanvasTop * multiplier);
+
+  return {
+    isOffCanvas,
+    cropParams: {
```

Comment:
<!-- propose improvements -->
**💬 suggestion:**

Should we perhaps instead add something like `getCropParams` callback inside this hook? Otherwise the calculation runs each time when rendering the right click menu, not just when actually needing to crop.

Maybe the same for `isOffCanvas`. Thoughts?

Reply:
Great idea

Reply:
Updated here:
https://github.com/GoogleForCreators/web-stories-wp/pull/12156/commits/4a9dafffc36e6b8670c0f47b205c877f7a1445ba